drivers/flash/flash_util.c

```diff
@@ -80,3 +87,199 @@ int z_impl_flash_flatten(const struct device *dev, off_t offset, size_t size)
+}
+
+int flash_region_crc(const struct device *dev, off_t offset, off_t size, uint32_t *crc,
+		     uint32_t *buf, size_t buf_size)
+{
+	off_t end;
+	uint32_t _buf[4];
+
+	if (!IS_ENABLED(CONFIG_CRC)) {
+		return -ENOTSUP;
+	}
+
+	if ((crc == NULL) || (offset < 0) || (size < 0)) {
+		return -EINVAL;
+	}
+
+	if (off_add_overflow(offset, size, &end)) {
+		return -EOVERFLOW;
+	}
+
+	if ((buf == NULL) || (buf_size == 0)) {
+		buf = _buf;
+		buf_size = sizeof(_buf);
+	}
+
+	if ((buf_size & (sizeof(uint32_t) - 1)) != 0) {
+		LOG_DBG("buffer size is not a multiple of uint32_t size");
+		return -EINVAL;
+	}
+
+	if (!device_is_ready(dev)) {
+		LOG_DBG("device not ready");
+		return -ENODEV;
+	}
+
+	for (uint32_t bytes = 0; size > 0; offset += bytes, size -= bytes) {
+		bytes = MIN(size, buf_size);
+		if (flash_read(dev, offset, buf, bytes) < 0) {
+			return -EIO;
+		}
+
+		*crc = crc32_ieee_update(*crc, (uint8_t *)buf, bytes);
+	}
+
+	return 0;
+}
+
+int flash_copy(const struct device *src_dev, off_t src_offset, const struct device *dst_dev,
+	       off_t dst_offset, off_t size, bool verify, uint32_t *buf, size_t buf_size)
+{
+	int ret;
+	uint32_t _buf[4];
+	uint32_t crc_src;
+	uint32_t crc_dst;
+	uint32_t write_size;
+
+	if ((buf == NULL) || (buf_size == 0)) {
+		buf = _buf;
+		buf_size = sizeof(_buf);
+	}
+
+	if (verify && !IS_ENABLED(CONFIG_CRC)) {
+		LOG_DBG("CRC support is not enabled");
+		return -ENOTSUP;
+	}
+
+	if ((src_offset < 0) || (dst_offset < 0) || (size < 0)) {
+		LOG_DBG("negative offset or size");
+		return -EINVAL;
+	}
+
+	if ((src_dev == dst_dev) && (src_offset == dst_offset)) {
+		/* nothing to do */
+		return 0;
+	}
+
+	if ((src_dev == dst_dev) &&
+	    (flash_regions_overlap(src_offset, size, dst_offset, size) != 0)) {
+		LOG_DBG("%s (%lx) and %s (%lx) overlap with size (%lx)", "source", (long)src_offset,
+			"destination", (long)dst_offset, (long)size);
+		return -EINVAL;
+	}
+
+	if (!device_is_ready(src_dev)) {
+		LOG_DBG("%s device not ready", "source");
+		return -ENODEV;
+	}
+
+	if (!device_is_ready(dst_dev)) {
+		LOG_DBG("%s device not ready", "destination");
+		return -ENODEV;
+	}
+
+	write_size = flash_get_write_block_size(dst_dev);
+	__ASSERT(sizeof(buf) >= write_size,
+		 "%s buf of size %u is too small for write_size of %u", __func__,
+		 (uint32_t)sizeof(buf), write_size);
+
+	if (verify && IS_ENABLED(CONFIG_CRC)) {
+		ret = flash_region_crc(src_dev, src_offset, size, &crc_src, buf, buf_size);
+		if (ret < 0) {
+			LOG_DBG("%s failed: %d", "flash_region_crc()", ret);
+			return ret;
+		}
+
+		ret = flash_region_crc(dst_dev, dst_offset, size, &crc_dst, buf, buf_size);
+		if (ret < 0) {
+			LOG_DBG("%s failed: %d", "flash_region_crc()", ret);
+			return ret;
+		}
+
+		if (crc_dst == crc_src) {
+			/* source and destination regions already have the same contents */
+			return 0;
+		}
+	}
+
+	/* This fails in testing for qemu_x86 iirc, with a flash simulator */
+	(void)flash_erase(dst_dev, dst_offset, size);
```

Comment:
The original implementation of this was using copy by erase size.

For some reason, I couldn't get the erase page size easily out of DT but I could revisit.

I think it would be fine to switch back to that.

Reply:
Fixed! (caller must erase prior to copy)